pkg/common/security/usergroup_ldap_resolver.go

```diff
@@ -0,0 +1,375 @@
+
+	"go.uber.org/zap"
+
+	"github.com/go-ldap/ldap/v3"
+
+	"github.com/apache/yunikorn-core/pkg/common"
+	"github.com/apache/yunikorn-core/pkg/log"
+)
+
+// This file contains the implementation of the LDAP resolver for user groups
+
+// LdapAccess defines the interface for LDAP operations
+type LdapAccess interface {
+	// DialURL establishes a connection to the LDAP server
+	DialURL(url string, options ...ldap.DialOpt) (*ldap.Conn, error)
+
+	// Bind authenticates with the LDAP server
+	Bind(conn *ldap.Conn, username, password string) error
+
+	// Search performs an LDAP search operation
+	Search(conn *ldap.Conn, searchRequest *ldap.SearchRequest) (*ldap.SearchResult, error)
+
+	// Close closes the LDAP connection
+	Close(conn *ldap.Conn)
+}
+
+// LdapAccessImpl implements the LdapAccess interface with real LDAP operations
+type LdapAccessImpl struct{}
+
+func (l *LdapAccessImpl) DialURL(url string, options ...ldap.DialOpt) (*ldap.Conn, error) {
+	return ldap.DialURL(url, options...)
+}
+
+func (l *LdapAccessImpl) Bind(conn *ldap.Conn, username, password string) error {
+	return conn.Bind(username, password)
+}
+
+func (l *LdapAccessImpl) Search(conn *ldap.Conn, searchRequest *ldap.SearchRequest) (*ldap.SearchResult, error) {
+	return conn.Search(searchRequest)
+}
+
+func (l *LdapAccessImpl) Close(conn *ldap.Conn) {
+	conn.Close()
+}
+
+// ldapAccessFactory is a function type that creates LdapAccess instances
+type ldapAccessFactory func(config *LdapResolverConfig) LdapAccess
+
+// defaultLdapAccessFactory is the default factory function that creates real LdapAccessImpl instances
+var defaultLdapAccessFactory ldapAccessFactory = func(config *LdapResolverConfig) LdapAccess {
+	return &LdapAccessImpl{}
+}
+
+// newLdapAccessImpl creates a new LdapAccess instance using the current factory
+// This can be replaced in tests to return mock implementations
+var newLdapAccessImpl = defaultLdapAccessFactory
+
+// resetLdapAccessFactory resets the factory to the default implementation
+// This is used in tests to ensure the global state is restored
+func resetLdapAccessFactory() {
+	newLdapAccessImpl = defaultLdapAccessFactory
+}
+
+// LDAPResolverConfig holds the configuration for the LDAP resolver
+type LdapResolverConfig struct {
+	Host         string
+	Port         int
+	BaseDN       string
+	Filter       string
+	GroupAttr    string
+	ReturnAttr   []string
+	BindUser     string
+	BindPassword string
+	Insecure     bool
+	SSL          bool
+}
+
+// Default values for the LDAP resolver
+var ldapConf = LdapResolverConfig{
+	Host:         common.DefaultLdapHost,
+	Port:         common.DefaultLdapPort,
+	BaseDN:       common.DefaultLdapBaseDN,
+	Filter:       common.DefaultLdapFilter,
+	GroupAttr:    common.DefaultLdapGroupAttr,
+	ReturnAttr:   common.DefaultLdapReturnAttr,
+	BindUser:     common.DefaultLdapBindUser,
+	BindPassword: common.DefaultLdapBindPassword,
+	Insecure:     common.DefaultLdapInsecure,
+	SSL:          common.DefaultLdapSSL,
+}
```

Comment:
We should consider moving this into the suggested new type `LdapLookup`. This is yet another global variable we should try to get rid of.